src/libs/tagging.c

```diff
@@ -130,24 +166,182 @@ static void update(dt_lib_module_t *self, int which)
+      for(GList *taglist = tags; taglist; taglist = g_list_next(taglist))
+      {
+        const gchar *tag = ((dt_tag_t *)taglist->data)->tag;
+        const guint id = ((dt_tag_t *)taglist->data)->id;
+        const guint tagc = ((dt_tag_t *)taglist->data)->count;
+        const guint sel = ((dt_tag_t *)taglist->data)->select;
+        if(tag == NULL) continue;
+        char **tokens;
+        tokens = g_strsplit(tag, "\1", -1);
+        if(tokens != NULL)
+        {
+          // find the number of common parts at the beginning of tokens and last_tokens
+          GtkTreeIter parent = last_parent;
+          const int tokens_length = string_array_length(tokens);
+          int common_length = 0;
+          if(last_tokens)
+          {
+            while(tokens[common_length] && last_tokens[common_length] &&
+                  !g_strcmp0(tokens[common_length], last_tokens[common_length]))
+            {
+              common_length++;
+            }
+
+            // point parent iter to where the entries should be added
+            for(int i = common_length; i < last_tokens_length; i++)
+            {
+              gtk_tree_model_iter_parent(model, &parent, &last_parent);
+              last_parent = parent;
+            }
+          }
+
+          // insert everything from tokens past the common part
+          char *pth = NULL;
+          for(int i = 0; i < common_length; i++)
+            pth = dt_util_dstrcat(pth, "%s|", tokens[i]);
+
+          for(char **token = &tokens[common_length]; *token; token++)
+          {
+            pth = dt_util_dstrcat(pth, "%s|", *token);
+            gchar *pth2 = g_strdup(pth);
+            pth2[strlen(pth2) - 1] = '\0';
+            gtk_tree_store_insert(GTK_TREE_STORE(model), &iter, common_length > 0 ? &parent : NULL, -1);
+            gtk_tree_store_set(GTK_TREE_STORE(model), &iter,
+                              DT_LIB_TAGGING_COL_TAG, *token,
+                              DT_LIB_TAGGING_COL_ID, (token == &tokens[tokens_length-1]) ? id : 0,
+                              DT_LIB_TAGGING_TREE_COL_PATH, pth2,
+                              DT_LIB_TAGGING_TREE_COL_COUNT, (token == &tokens[tokens_length-1]) ? tagc : 0,
+                              DT_LIB_TAGGING_TREE_COL_SEL, sel,
+                              -1);
+            if (sel)
+            { // propagate selection to parents
+              guint sel1;
+              GtkTreeIter parent2, child = iter;
+              while(gtk_tree_model_iter_parent(model, &parent2, &child))
+              {
+                gtk_tree_model_get(model, &parent2, DT_LIB_TAGGING_TREE_COL_SEL, &sel1, -1);
+                if (!sel1)
+                  gtk_tree_store_set(GTK_TREE_STORE(model), &parent2, DT_LIB_TAGGING_TREE_COL_SEL, 1, -1);
+                child = parent2;
+              }
+            }
+            common_length++;
+            parent = iter;
+            g_free(pth2);
+          }
+          g_free(pth);
```

Comment:
and replace this with:
g_strfreev (tokens);

Less error prone and when reading code it makes it clear that the vector is freed.